master/internal/api_master.go

```diff
@@ -25,3 +26,60 @@ func (a *apiServer) GetMaster(
 		Config:    configStruct,
 	}, err
 }
+
+// effectiveOffset Returns effective offset.
+func effectiveOffset(reqOffset int, totalItems int) (offset int) {
+	offset = reqOffset
+	if reqOffset < 0 {
+		offset = totalItems + offset
+		if offset < 0 {
+			offset = 0
+		}
+	}
+	return offset
+}
+
+// effectiveLimit Returns effective limit.
+// Input: Limit 0 is treated as no limit
+// Output: Limit -1 is treated as no limit
+func effectiveLimit(reqLimit int, offset int, totalItems int) (limit int) {
+	limit = -1
+	if reqLimit != 0 {
+		limit = reqLimit
+		if limit > totalItems-offset {
+			limit = totalItems - offset
+		}
+	}
+	return limit
+}
+
+// TODO could we have this work with generic requests that have offset and limit?
+func effectiveOffsetNLimit(reqOffset int, reqLimit int, totalItems int) (offset int, limit int) {
+	offset = effectiveOffset(reqOffset, totalItems)
+	limit = effectiveLimit(reqLimit, offset, totalItems)
+	return offset, limit
+}
+
+func (a *apiServer) MasterLogs(
+	req *apiv1.MasterLogsRequest, resp apiv1.Determined_MasterLogsServer) error {
+	if err := grpc.ValidateRequest(
+		grpc.ValidateLimit(req.Limit),
+	); err != nil {
+		return err
+	}
+	total := a.m.logs.Len()
+	offset, limit := effectiveOffsetNLimit(int(req.Offset), int(req.Limit), total)
+
+	for {
+		for _, log := range a.m.logs.Entries(offset, -1, limit) {
```

Comment:
@stoksc wouldn't this pin the CPU if there are logs to report and the user has req.Follow set?

Reply:
Yep, looks like it would to me, nice catch. We could add a tiny timeout if it grabs a slice of entries without any logs (there are other things I can think of doing, but they're an order of magnitude more work)?

Reply:
alright that sounds good for now. I rather reserve more work for another PR so we can wrap up Yoni's PR without much change.